dart/common/detail/Addon.h

```diff
@@ -0,0 +1,338 @@
+/*
+ * Copyright (c) 2015, Georgia Tech Research Corporation
+ * All rights reserved.
+ *
+ * Author(s): Michael X. Grey <mxgrey@gatech.edu>
+ *
+ * Georgia Tech Graphics Lab and Humanoid Robotics Lab
+ *
+ * Directed by Prof. C. Karen Liu and Prof. Mike Stilman
+ * <karenliu@cc.gatech.edu> <mstilman@cc.gatech.edu>
+ *
+ * This file is provided under the following "BSD-style" License:
+ *   Redistribution and use in source and binary forms, with or
+ *   without modification, are permitted provided that the following
+ *   conditions are met:
+ *   * Redistributions of source code must retain the above copyright
+ *     notice, this list of conditions and the following disclaimer.
+ *   * Redistributions in binary form must reproduce the above
+ *     copyright notice, this list of conditions and the following
+ *     disclaimer in the documentation and/or other materials provided
+ *     with the distribution.
+ *   THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND
+ *   CONTRIBUTORS "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES,
+ *   INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF
+ *   MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ *   DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR
+ *   CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ *   SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ *   LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF
+ *   USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED
+ *   AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ *   LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+ *   ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ *   POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#ifndef DART_COMMON_DETAIL_ADDON_H_
+#define DART_COMMON_DETAIL_ADDON_H_
+
+#include <cassert>
+
+#include "dart/common/Console.h"
+#include "dart/common/Addon.h"
+
+namespace dart {
+namespace common {
+
+#define DART_COMMON_CAST_NEW_MANAGER_TYPE(Base, ManagerType, newManager,\
+                                          castedManager, func)\
+  ManagerType* castedManager = dynamic_cast<ManagerType*>(newManager);\
+  if(nullptr == castedManager)\
+  {\
+    dterr << "[" << typeid(Base).name() << "::" << #func << "] Attempting to "\
+          << "use a [" << typeid(newManager).name() << "] type manager, but "\
+          << "this Addon is only designed to be attached to a ["\
+          << typeid(ManagerType).name() << "] type manager. This may cause "\
+          << "undefined behavior!\n";\
+    assert(false);\
+  }
+
+#define DART_COMMON_CAST_NEW_MANAGER_TYPE_AND_RETURN_NULL_IF_BAD(\
+  Base, ManagerType, newManager, castedManager, func)\
+  DART_COMMON_CAST_NEW_MANAGER_TYPE(Base, ManagerType, newManager, castedManager, func)\
+  if(nullptr == castedManager) return nullptr;
```

Comment:
I don't see where a segfault would be created. Are you referring to the abort caused by ``assert(false)`` on line 58? I believe most (probably all?) compilers will wash away calls to ``assert()`` when compiling in release mode, so it's still useful to make sure we return a ``nullptr`` in case this issue arises on code compiled in release mode.